stdlib/format.ml

```diff
@@ -441,80 +388,62 @@ let format_pp_token state size = function
       | Pp_vbox -> break_new_line state off width
       | Pp_hbox -> break_same_line state n
       end
-    | [] -> () (* No open box. *)
+    | None -> () (* No open box. *)
     end
 
    | Pp_open_tag tag_name ->
      let marker = state.pp_mark_open_tag tag_name in
      pp_output_string state marker;
-     state.pp_mark_stack <- tag_name :: state.pp_mark_stack
+     Stack.push tag_name state.pp_mark_stack
 
    | Pp_close_tag ->
-     begin match state.pp_mark_stack with
-     | tag_name :: tags ->
+     begin match Stack.pop_opt state.pp_mark_stack with
+     | Some tag_name ->
        let marker = state.pp_mark_close_tag tag_name in
        pp_output_string state marker;
-       state.pp_mark_stack <- tags
-     | [] -> () (* No more tag to close. *)
+     | None -> () (* No more tag to close. *)
      end
 
 
 (* Print if token size is known else printing is delayed.
-   Size is known when not negative.
    Printing is delayed when the text waiting in the queue requires
-   more room to format than exists on the current line.
-
-   Note: [advance_loop] must be tail recursive to prevent stack overflows. *)
-let rec advance_loop state =
-  match peek_queue state.pp_queue with
-  | {elem_size = size; token = tok; length = len} ->
-    let size = int_of_size size in
-    if not
-         (size < 0 &&
-          (state.pp_right_total - state.pp_left_total < state.pp_space_left))
-    then begin
-      ignore (take_queue state.pp_queue);
-      format_pp_token state (if size < 0 then pp_infinity else size) tok;
-      state.pp_left_total <- len + state.pp_left_total;
-      advance_loop state
+   more room to format than exists on the current line. *)
+let rec advance_left state =
+  match Queue.peek_opt state.pp_queue with
+  | None -> () (* No tokens to print *)
+  | Some { size; token; length } ->
+    let is_size_known = Size.is_known size in
```

Comment:
Not convinced that the `is_size_known` binding makes the code more readable, it's mostly transparent.